src/StreamJsonRpc/ProxyGeneration.cs

```diff
@@ -0,0 +1,353 @@
+
+                    // public event EventHandler EventName;
+                    var evtBuilder = proxyTypeBuilder.DefineEvent(evt.Name, evt.Attributes, evt.EventHandlerType);
+
+                    // private EventHandler eventName;
+                    var evtField = proxyTypeBuilder.DefineField(evt.Name, evt.EventHandlerType, FieldAttributes.Private);
+
+                    // add_EventName
+                    var addRemoveHandlerParams = new Type[] { evt.EventHandlerType };
+                    const MethodAttributes methodAttributes = MethodAttributes.Final | MethodAttributes.HideBySig | MethodAttributes.Public | MethodAttributes.SpecialName | MethodAttributes.NewSlot | MethodAttributes.Virtual;
+                    var addMethod = proxyTypeBuilder.DefineMethod($"add_{evt.Name}", methodAttributes, null, addRemoveHandlerParams);
+                    ImplementEventAccessor(addMethod.GetILGenerator(), evtField, DelegateCombineMethod);
+                    evtBuilder.SetAddOnMethod(addMethod);
+
+                    // remove_EventName
+                    var removeMethod = proxyTypeBuilder.DefineMethod($"remove_{evt.Name}", methodAttributes, null, addRemoveHandlerParams);
+                    ImplementEventAccessor(removeMethod.GetILGenerator(), evtField, DelegateRemoveMethod);
+                    evtBuilder.SetRemoveOnMethod(removeMethod);
+
+                    // void OnEventName(EventArgs args)
+                    var eventArgsType = evt.EventHandlerType.GetTypeInfo().GetDeclaredMethod(nameof(EventHandler.Invoke)).GetParameters()[1].ParameterType;
+                    var raiseEventMethod = proxyTypeBuilder.DefineMethod(
+                        $"On{evt.Name}",
+                        MethodAttributes.HideBySig | MethodAttributes.Private,
+                        null,
+                        new Type[] { eventArgsType });
+                    ImplementRaiseEventMethod(raiseEventMethod.GetILGenerator(), evtField, jsonRpcField);
+
+                    ctorActions.Add(new Action<ILGenerator>(il =>
+                    {
+                        var addLocalRpcMethod = typeof(JsonRpc).GetRuntimeMethod(nameof(JsonRpc.AddLocalRpcMethod), new Type[] { typeof(string), typeof(Delegate) });
+                        var delegateCtor = typeof(Action<>).MakeGenericType(eventArgsType).GetTypeInfo().DeclaredConstructors.Single();
+
+                        // rpc.AddLocalRpcMethod("EventName", new Action<EventArgs>(this.OnEventName));
+                        il.Emit(OpCodes.Ldarg_1); // .ctor's rpc parameter
+                        il.Emit(OpCodes.Ldstr, evt.Name);
+                        il.Emit(OpCodes.Ldarg_0);
+                        il.Emit(OpCodes.Ldftn, raiseEventMethod);
+                        il.Emit(OpCodes.Newobj, delegateCtor);
+                        il.Emit(OpCodes.Callvirt, addLocalRpcMethod);
+                    }));
+                }
+
+                // .ctor(JsonRpc)
+                {
```

Comment:
Is there a reason beyond organization to introduce a new scope for this?

Reply:
Besides readability, it allows me to redefine locals for each generated method. Sharing locals between generated methods tends to make for very brittle code.